File: c_plus_lib/netlist_graph/libsrc/edge.cpp
#include "edge.hpp"

#include <sstream>
#include <vector>

#include "node.hpp"

edge::edge( void )
{
	a_id = -1;				// Refers to node (instance of a component)
	a_name = "";			// Refers to node name (instance name of a component)
	a_pad_id = -1;
	a_pad_name = "";
	a_size_x = 0.0;
	a_size_y = 0.0;
	a_pos_x = 0.0;
	a_pos_y = 0.0;
	a_isPlaced = false;

	b_id = -1;
	b_name = "";
	b_pad_id = -1;
	b_pad_name = "";
	b_size_x = 0.0;
	b_size_y = 0.0;
	b_pos_x = 0.0;
	b_pos_y = 0.0;
	b_isPlaced = false;

	net_id = -1;
	net_name = "";
	//common_net_points = 0;
	//uncommon_net_points = 0;

	power_rail = 0;
}

int8_t edge::create_from_string_short( std::string s )
{
	std::stringstream ss(s);
	std::vector<std::string> fields;
	std::string field;

	while(ss.good())
	{
		getline(ss,field,',');
		fields.push_back(field);
	}

	a_id = stoi(fields[0]);
	a_size_x = stod(fields[1]);
	a_size_y = stod(fields[2]);
	a_pos_x = stod(fields[3]);
	a_pos_y = stod(fields[4]);
	a_isPlaced = (bool) stoi(fields[5]);

	b_id = stoi(fields[6]);
	b_size_x = stod(fields[7]);
	b_size_y = stod(fields[8]);
	b_pos_x = stod(fields[9]);
	b_pos_y = stod(fields[10]);
	b_isPlaced = (bool) stoi(fields[11]);

	net_id = stoi(fields[12]);

	//common_net_points = stoi(fields[13]);
	//uncommon_net_points = stoi(fields[14]);

	power_rail = stoi(fields[13]);

	return 0;
}

int8_t edge::create_from_string_long( std::string s )
{
	std::stringstream ss(s);
	std::vector<std::string> fields;
	std::string field;

	while(ss.good())
	{
		getline(ss,field,',');
		fields.push_back(field);
	}

	a_id = stoi(fields[0]);
	a_pad_id = stoi(fields[1]);
	a_pad_name = fields[2];

	a_size_x = stod(fields[3]);
	a_size_y = stod(fields[4]);
	a_pos_x = stod(fields[5]);
	a_pos_y = stod(fields[6]);
	a_isPlaced = (bool) stoi(fields[7]);

	b_id = stoi(fields[8]);
	b_pad_id = stoi(fields[9]);
	b_pad_name = fields[10];

	b_size_x = stod(fields[11]);
	b_size_y = stod(fields[12]);
	b_pos_x = stod(fields[13]);
	b_pos_y = stod(fields[14]);
	b_isPlaced = (bool) stoi(fields[15]);

	net_id = stoi(fields[16]);
	net_name = fields[17];
	//common_net_points = stoi(fields[18]);
	//uncommon_net_points = stoi(fields[19]);

	power_rail = stoi(fields[18]);

	return 0;
}

std::pair<double, double> edge::get_size(int id)
{
	std::pair<double,double> p;

	if (id == 0) p = std::make_pair(a_size_x, a_size_y);
	else p = std::make_pair(b_size_x, b_size_y);

	return p;
}

int edge::set_size(int id, std::pair<double, double> p)
{
	if (id == 0) { a_size_x = p.first; a_size_y = p.second; }
	else { b_size_x = p.first; b_size_y = p.second; }

	return 0;
}

std::pair<double, double> edge::get_pos(int id)
{
	std::pair<double,double> p;

	if (id == 0) p = std::make_pair(a_pos_x, a_pos_y);
	else p = std::make_pair(b_pos_x, b_pos_y);

	return p;
}

int edge::set_pos(int id, std::pair<double, double> p)
{
	if (id == 0) { a_pos_x = p.first; a_pos_y = p.second; }
	else { b_pos_x = p.first; b_pos_y = p.second; }

	return 0;
}

void edge::print_to_console( bool format )
{

    std::cout << a_id << ",";                                               // instance id
    if(format)
    {
            std::cout << a_pad_id << ","            // pad id
                    << a_pad_name << ",";   // pad name

    }
    std::cout << a_size_x << ","
                    << a_size_y << ","
                    << a_pos_x << ","
                    << a_pos_y << ","
                    << a_isPlaced << ","
                    //<< n1 << ", "
                    << b_id << ",";
    if(format)
    {
            std::cout << b_pad_id << ","
                    << b_pad_name<< ",";

    }
    std::cout << b_size_x << ","
                    << b_size_y << ","
                    << b_pos_x << ","
                    << b_pos_y << ","
                    << b_isPlaced << ","
                    //<< n2 << ", "
                    << net_id << ",";
    if(format) std::cout << net_name << ",";
    //std::cout << common_net_points << ","
    //              << uncommon_net_points << ",";
    std::cout << power_rail
                    << std::endl;

}

int edge::print( bool print_csv )
{
	if(print_csv)
	{
		if (a_name == "") std::cout << a_id << ",";
		else std::cout << a_name << ",";
		std::cout << a_size_x << "," << a_size_y << ",";
		std::cout << a_pos_x << "," << a_pos_y << ",";

		if (b_name == "") std::cout << b_id << ",";
		else std::cout << b_name << ",";
		std::cout << b_size_x << "," << b_size_y << ",";
		std::cout << b_pos_x << "," << b_pos_y << ",";
		std::cout << std::endl;
	}
	else
	{
		std::cout << "Edge ";

		std::cout << a_id;
		if (a_name != "") std::cout << "(" << a_name << ")";
		std::cout << " - ";
		std::cout << b_id;
		if (b_name != "") std::cout << "(" << b_name << ")";
		std::cout << std::endl;

		std::cout << "  Size        : (" << a_size_x << "," << a_size_y << ") - (" << b_size_x << "," << b_size_y << ")" << std::endl;
		std::cout << "  Position    : (" << a_pos_x << "," << a_pos_y << ") - (" << b_pos_x << "," << b_pos_y << ")" << std::endl;
	    std::cout << std::endl;
	}



	return 0;
}


int edge::format_string_long( std::string& line )
{
	std::stringstream l;

	l << a_id << ","
		<< a_pad_id << ","
		<< a_pad_name << ","
		<< a_size_x << ","
		<< a_size_y << ","
		<< a_pos_x << ","
		<< a_pos_y << ","
		<< a_isPlaced << ","
		<< b_id << ","
		<< b_pad_id << ","
		<< b_pad_name << ","
		<< b_size_x << ","
		<< b_size_y << ","
		<< b_pos_x << ","
		<< b_pos_y << ","
		<< b_isPlaced << ","
		<< net_id << ","
		<< net_name << ","
		<< power_rail
		<< std::endl;

	line = l.str();
	return 0;
}

File: c_plus_lib/pcb_python/netlist/edge.py
"""
Edge类 - 表示电路图中的边（连接）
对应C++的edge.hpp
"""

from typing import Tuple, Optional
from .utils import Utils

class Edge:
    """表示电路图中的边（连接）"""
    
    def __init__(self):
        # 节点A的信息
        self._a_id = -1
        self._a_name = ""
        self._a_pad_id = -1
        self._a_pad_name = ""
        self._a_size_x = 0.0
        self._a_size_y = 0.0
        self._a_pos_x = 0.0
        self._a_pos_y = 0.0
        self._a_is_placed = False
        
        # 节点B的信息
        self._b_id = -1
        self._b_name = ""
        self._b_pad_id = -1
        self._b_pad_name = ""
        self._b_size_x = 0.0
        self._b_size_y = 0.0
        self._b_pos_x = 0.0
        self._b_pos_y = 0.0
        self._b_is_placed = False
        
        # 网络信息
        self._net_id = -1
        self._net_name = ""
        self._power_rail = 0
    
    def create_from_string_short(self, s: str) -> int:
        """从短格式字符串创建边"""
        try:
            fields = Utils.parse_csv_line(s)
            if len(fields) < 20:
                return -1
            
            # 节点A信息
            self._a_id = Utils.safe_int(fields[0])
            self._a_name = fields[1]
            self._a_pad_id = Utils.safe_int(fields[2])
            self._a_size_x = Utils.safe_float(fields[3])
            self._a_size_y = Utils.safe_float(fields[4])
            self._a_pos_x = Utils.safe_float(fields[5])
            self._a_pos_y = Utils.safe_float(fields[6])
            self._a_is_placed = Utils.safe_int(fields[7]) == 1
            
            # 节点B信息
            self._b_id = Utils.safe_int(fields[8])
            self._b_name = fields[9]
            self._b_pad_id = Utils.safe_int(fields[10])
            self._b_size_x = Utils.safe_float(fields[11])
            self._b_size_y = Utils.safe_float(fields[12])
            self._b_pos_x = Utils.safe_float(fields[13])
            self._b_pos_y = Utils.safe_float(fields[14])
            self._b_is_placed = Utils.safe_int(fields[15]) == 1
            
            # 网络信息
            self._net_id = Utils.safe_int(fields[16])
            self._net_name = fields[17].strip('"') if len(fields) > 17 else ""
            self._power_rail = Utils.safe_int(fields[18]) if len(fields) > 18 else 0
            
            return 0
        except Exception:
            return -1
    
    def create_from_string_long(self, s: str) -> int:
        """从长格式字符串创建边"""
        try:
            fields = Utils.parse_csv_line(s)
            
            # 检查字段数量，支持19字段的标准格式
            if len(fields) < 19:
                return -1
            
            # 节点A信息
            self._a_id = Utils.safe_int(fields[0])
            self._a_pad_id = Utils.safe_int(fields[1])
            self._a_pad_name = fields[2]
            self._a_size_x = Utils.safe_float(fields[3])
            self._a_size_y = Utils.safe_float(fields[4])
            self._a_pos_x = Utils.safe_float(fields[5])
            self._a_pos_y = Utils.safe_float(fields[6])
            self._a_is_placed = Utils.safe_int(fields[7]) == 1
            
            # 节点B信息
            self._b_id = Utils.safe_int(fields[8])
            self._b_pad_id = Utils.safe_int(fields[9])
            self._b_pad_name = fields[10]
            self._b_size_x = Utils.safe_float(fields[11])
            self._b_size_y = Utils.safe_float(fields[12])
            self._b_pos_x = Utils.safe_float(fields[13])
            self._b_pos_y = Utils.safe_float(fields[14])
            self._b_is_placed = Utils.safe_int(fields[15]) == 1
            
            # 网络信息
            self._net_id = Utils.safe_int(fields[16])
            self._net_name = fields[17].strip('"') if len(fields) > 17 else ""
            self._power_rail = Utils.safe_int(fields[18]) if len(fields) > 18 else 0
            
            # 设置节点名称（从pad_name推断）
            self._a_name = f"Node_{self._a_id}"
            self._b_name = f"Node_{self._b_id}"
            
            return 0
        except Exception:
            return -1
    
    def get_net_id(self) -> int:
        """获取网络ID"""
        return self._net_id
    
    def get_net_name(self) -> str:
        """获取网络名称"""
        return self._net_name
    
    def get_power_rail(self) -> int:
        """获取电源轨信息"""
        return self._power_rail
    
    def get_instance_id(self, node: int) -> int:
        """获取指定节点的实例ID"""
        return self._b_id if node else self._a_id
    
    def get_instance_is_placed(self, node: int) -> bool:
        """获取指定节点的放置状态"""
        return self._b_is_placed if node else self._a_is_placed
    
    def get_pad_name(self, node: int) -> str:
        """获取指定节点的焊盘名称"""
        return self._b_pad_name if node else self._a_pad_name
    
    def get_pad_id(self, node: int) -> int:
        """获取指定节点的焊盘ID"""
        return self._b_pad_id if node else self._a_pad_id
    
    def get_edge_connectivity(self) -> Tuple[int, int]:
        """获取边的连接信息"""
        return (self._a_id, self._b_id)
    
    def get_size(self, node_id: int) -> Tuple[float, float]:
        """获取指定节点的尺寸"""
        if node_id == self._a_id:
            return (self._a_size_x, self._a_size_y)
        elif node_id == self._b_id:
            return (self._b_size_x, self._b_size_y)
        else:
            return (0.0, 0.0)
    
    def set_size(self, node_id: int, size: Tuple[float, float]) -> int:
        """设置指定节点的尺寸"""
        if node_id == self._a_id:
            self._a_size_x, self._a_size_y = size
        elif node_id == self._b_id:
            self._b_size_x, self._b_size_y = size
        else:
            return -1
        return 0
    
    def get_pos(self, node_id: int) -> Tuple[float, float]:
        """获取指定节点的位置"""
        if node_id == self._a_id:
            return (self._a_pos_x, self._a_pos_y)
        elif node_id == self._b_id:
            return (self._b_pos_x, self._b_pos_y)
        else:
            return (0.0, 0.0)
    
    def set_pos(self, node_id: int, pos: Tuple[float, float]) -> int:
        """设置指定节点的位置"""
        if node_id == self._a_id:
            self._a_pos_x, self._a_pos_y = pos
        elif node_id == self._b_id:
            self._b_pos_x, self._b_pos_y = pos
        else:
            return -1
        return 0
    
    def print_to_console(self, format_long: bool) -> None:
        """打印边信息到控制台"""
        if format_long:
            print(f"Edge: {self._a_name}({self._a_id}) -> {self._b_name}({self._b_id}), "
                  f"Net: {self._net_name}({self._net_id}), Power: {self._power_rail}")
        else:
            print(f"{self._a_id},{self._a_name},{self._a_pad_id},{self._a_size_x:.6f},{self._a_size_y:.6f},"
                  f"{self._a_pos_x:.6f},{self._a_pos_y:.6f},{1 if self._a_is_placed else 0},"
                  f"{self._b_id},{self._b_name},{self._b_pad_id},{self._b_size_x:.6f},{self._b_size_y:.6f},"
                  f"{self._b_pos_x:.6f},{self._b_pos_y:.6f},{1 if self._b_is_placed else 0},"
                  f"{self._net_id},\"{self._net_name}\",{self._power_rail}")
    
    def print(self, print_csv: bool) -> int:
        """打印边信息"""
        self.print_to_console(not print_csv)
        return 0
    
    def format_string_long(self) -> str:
        """格式化长字符串"""
        line = f"{self._a_id},{self._a_name},{self._a_pad_id},{self._a_size_x:.6f},{self._a_size_y:.6f},"
        line += f"{self._a_pos_x:.6f},{self._a_pos_y:.6f},{1 if self._a_is_placed else 0},"
        line += f"{self._b_id},{self._b_name},{self._b_pad_id},{self._b_size_x:.6f},{self._b_size_y:.6f},"
        line += f"{self._b_pos_x:.6f},{self._b_pos_y:.6f},{1 if self._b_is_placed else 0},"
        line += f"{self._net_id},\"{self._net_name}\",{self._power_rail}"
        return line
    
    # 设置器方法
    def set_id(self, node: int, node_id: int) -> None:
        """设置指定节点的ID"""
        if node:
            self._b_id = node_id
        else:
            self._a_id = node_id
    
    def set_name(self, node: int, name: str) -> None:
        """设置指定节点的名称"""
        if node:
            self._b_name = name
        else:
            self._a_name = name
    
    def set_pad_id(self, node: int, pad_id: int) -> None:
        """设置指定节点的焊盘ID"""
        if node:
            self._b_pad_id = pad_id
        else:
            self._a_pad_id = pad_id
    
    def set_pad_name(self, node: int, pad_name: str) -> None:
        """设置指定节点的焊盘名称"""
        if node:
            self._b_pad_name = pad_name
        else:
            self._a_pad_name = pad_name
    
    def set_is_placed(self, node: int, is_placed: bool) -> None:
        """设置指定节点的放置状态"""
        if node:
            self._b_is_placed = is_placed
        else:
            self._a_is_placed = is_placed
    
    def set_net_id(self, net_id: int) -> None:
        """设置网络ID"""
        self._net_id = net_id
    
    def set_net_name(self, net_name: str) -> None:
        """设置网络名称"""
        self._net_name = net_name
    
    def set_power_rail(self, power_rail: int) -> None:
        """设置电源轨信息"""
        self._power_rail = power_rail
    
    def __str__(self) -> str:
        """字符串表示"""
        return f"Edge({self._a_name}({self._a_id}) -> {self._b_name}({self._b_id}), Net: {self._net_name})"
    
    def __repr__(self) -> str:
        """详细字符串表示"""
        return self.__str__() 
    def get_node_id_a(self):
     return self._a_id

    def get_node_id_b(self):
     return self._b_id